blockindex/liquidstaking_indexer.go

```diff
@@ -0,0 +1,950 @@
+	if err != nil {
+		if !errors.Is(err, db.ErrNotExist) {
+			return err
+		}
+		height = 0
+	} else {
+		height = deserializeUint64(h)
+	}
+	s.cleanCache.putHeight(height)
+
+	// load bucket info
+	ks, vs, err := s.kvstore.Filter(_liquidStakingBucketInfoNS, func(k, v []byte) bool { return true }, nil, nil)
+	if err != nil {
+		if !errors.Is(err, db.ErrBucketNotExist) {
+			return err
+		}
+	}
+	for i := range vs {
+		var b BucketInfo
+		if err := b.deserialize(vs[i]); err != nil {
+			return err
+		}
+		s.cleanCache.putBucketInfo(deserializeUint64(ks[i]), &b)
+	}
+
+	// load bucket type
+	ks, vs, err = s.kvstore.Filter(_liquidStakingBucketTypeNS, func(k, v []byte) bool { return true }, nil, nil)
+	if err != nil {
+		if !errors.Is(err, db.ErrBucketNotExist) {
+			return err
+		}
+	}
+	for i := range vs {
+		var b BucketType
+		if err := b.deserialize(vs[i]); err != nil {
+			return err
+		}
+		s.cleanCache.putBucketType(deserializeUint64(ks[i]), &b)
+	}
+	return nil
+}
+
+func (s *liquidStakingIndexer) getBucketTypeIndex(amount *big.Int, duration time.Duration) (uint64, bool) {
+	id, ok := s.dirtyCache.getBucketTypeIndex(amount, duration)
+	if ok {
+		return id, true
+	}
+	id, ok = s.cleanCache.getBucketTypeIndex(amount, duration)
+	return id, ok
+}
+
+func (s *liquidStakingIndexer) getBucketTypeCount() uint64 {
+	base := len(s.cleanCache.idBucketTypeMap)
+	add := 0
+	for k, dbt := range s.dirtyCache.idBucketTypeMap {
+		_, ok := s.cleanCache.idBucketTypeMap[k]
+		if dbt != nil && !ok {
+			add++
+		} else if dbt == nil && ok {
+			add--
+		}
+	}
+	return uint64(base + add)
+}
+
+func (s *liquidStakingIndexer) getBucketType(id uint64) (*BucketType, bool) {
+	bt, ok := s.dirtyCache.getBucketType(id)
+	if ok {
+		return bt, true
+	}
+	bt, ok = s.cleanCache.getBucketType(id)
+	return bt, ok
+}
+
+func (s *liquidStakingIndexer) putHeight(h uint64) {
+	s.dirty.Put(_liquidStakingHeightNS, _liquidStakingHeightKey, serializeUint64(h), "failed to put height")
+	s.dirtyCache.putHeight(h)
+}
+
+func (s *liquidStakingIndexer) putBucketType(id uint64, bt *BucketType) {
+	s.dirty.Put(_liquidStakingBucketTypeNS, serializeUint64(id), bt.serialize(), "failed to put bucket type")
+	s.dirtyCache.putBucketType(id, bt)
+}
+
+func (s *liquidStakingIndexer) putBucketInfo(id uint64, bi *BucketInfo) {
+	s.dirty.Put(_liquidStakingBucketInfoNS, serializeUint64(id), bi.serialize(), "failed to put bucket info")
+	s.dirtyCache.putBucketInfo(id, bi)
+}
+
+func (s *liquidStakingIndexer) getBucketInfo(id uint64) (*BucketInfo, bool) {
+	bi, ok := s.dirtyCache.getBucketInfo(id)
+	if ok {
+		return bi, bi != nil
+	}
+	bi, ok = s.cleanCache.getBucketInfo(id)
+	return bi, ok
+}
+
+func (s *liquidStakingIndexer) burnBucket(id uint64) {
+	s.dirty.Delete(_liquidStakingBucketInfoNS, serializeUint64(id), "failed to delete bucket info")
+	s.dirtyCache.markDeleteBucketInfo(id)
+}
+
+func (s *liquidStakingIndexer) commit() error {
+	if err := s.cleanCache.writeBatch(s.dirty); err != nil {
```

Comment:
i don't think we need 2 caches, one clean and one dirty? just need to add all the changes into the local map when `commit` those changes to db
let's chat to clarify

Reply:
see if possible to do `cleanCache.merge(dirtyCache)`